projects/find-waffle/src/libs/keyboard/KeyMap.ts

```diff
@@ -0,0 +1,147 @@
+import { KeyBindingValidationError } from './errors';
+import {
+  extractModifiersFromEvent,
+  findModifier,
+  parseKeyBinding,
+} from './utils';
+
+export const NONE = 0b0000;
+export const SHIFT = 0b0001;
+export const CTRL = 0b0010;
+export const ALT = 0b0100;
+export const META = 0b1000;
+
+type Callback = () => void;
+type Modifier = number;
+type Code = string;
+type BindingTable = Map<Modifier, Map<Code, [Callback?, Callback?]>>;
+type ReleaseCallbackTable = Map<Code, Callback>;
+
+export class KeyMap {
+  private static _currentProfile?: KeyMap;
+  private static readonly _pressedKeys: Set<string> = new Set();
+
+  private readonly bindingTable: BindingTable;
+  private readonly releaseCallbackTable: ReleaseCallbackTable;
+
+  static {
+    window.addEventListener('keydown', (event) => {
+      KeyMap._pressedKeys.add(event.code);
+      KeyMap._currentProfile?.handleKeyDown(event);
+    });
+
+    window.addEventListener('keyup', (event) => {
+      KeyMap._pressedKeys.delete(event.code);
+      KeyMap._currentProfile?.handleKeyUp(event);
+    });
+  }
+
+  constructor() {
+    this.bindingTable = new Map();
+    this.releaseCallbackTable = new Map();
+  }
+
+  get pressedKeys() {
+    return new Set(KeyMap._pressedKeys);
+  }
+
+  get isActivated() {
+    return KeyMap._currentProfile === this;
+  }
+
+  bind(
+    keyBinding: string | string[],
+    pressCallback?: () => void,
+    releaseCallback?: () => void,
+  ): ThisType<KeyMap> {
+    // key binding 이 여러개 일 경우
+    if (Array.isArray(keyBinding)) {
+      for (const binding of keyBinding) {
+        this.bind(binding, pressCallback, releaseCallback);
+      }
+      return this;
+    }
+
+    try {
+      const { modifiers, code } = parseKeyBinding(keyBinding);
+
+      const codeModifier = findModifier(
+        code.match(/^(Shift|Control|Alt|Meta)/)?.[1] ?? '',
+      );
+
+      this.addKeyBinding(
+        (codeModifier ?? modifiers) | modifiers, // == modifiers 또는 (codeModifier | modifiers)
```

Comment:
`findModifier`에서 `undefined` 대신에 `NONE`을 반환하면 아래처럼 바꿀 수 있을 것 같은데 어떤가요? `extractModifiersFromEvent`에서도 눌린 키가 없을 때 `NONE`을 사용하고 있어서 통일성있을 것 같아요.

```suggestion
      this.addKeyBinding(
        codeModifier | modifiers,
```

Reply:
ad953a3
천재